File: Baekjoon/DP/1699.py
# https://www.acmicpc.net/problem/1699

# https://www.acmicpc.net/problem/1699

import sys
input = sys.stdin.readline

n = int(input())

dp = [i for i in range(n + 1)]

for i in range(2, int(n ** 0.5) + 1):
    for j in range(i * i, n + 1):
        dp[j] = min(dp[j], dp[j - i * i] + 1)

print(dp[n])

File: Baekjoon/DP/1699.cpp
// https://www.acmicpc.net/problem/1699

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n;
    cin >> n;
    vector<int> d(n + 1);
    for (int i = 1; i <= n; i++) {
        d[i] = i;
        for (int j = 1; j * j <= i; j++) {
            if (d[i] > d[i - j * j] + 1) {
                d[i] = d[i - j * j] + 1;
            }
        }
    }
    cout << d[n] << '\n';
    return 0;
}